### PR TITLE
refactor(ui): Extract magic numbers from Dialog

### PR DESCRIPTION
**Pull request by tibetiroka**
**Refactor**

This PR replaces Dialog's magic numbers with named (and explained) values.

## Summary
- Moves all padding and margin values to the header file as static constants.
- Replaces all width and height values with the actual sprites' width and height
- Fixes button hitboxes to match the buttons' visual size exactly
- Fixes an edge case when re-slimming wide dialogs (it previously ignored the padding, only counted the margin)

## Screenshots
If you can't tell what's before and what's after, it works.
image
image

## Usage examples
Readable code? Dialog sprites we can actually edit now?

## Testing Done
See the screenshots.

### CODE REVIEW

**Comment by tibetiroka** [comment]
Wrong :)

**Comment by warp-core** [comment]
Why private, static, compile-time constant fields in the header file and not compile-time constants in the Dialog.cpp anonymous namespace?

**Comment by tibetiroka** [comment]
> Why private, static, compile-time constant fields in the header file and not compile-time constants in the Dialog.cpp anonymous namespace?

Ideally, they would be defined in interfaces anway, especially since input fields use the same values in conversations... 

Sure, I could move them over to the C++ file. I thought they were properties of the dialog class itself, since most of these refer to the sprites' details (such as margins), and therefore aren't implementation details.

File: source/Dialog.cpp [hunk 1]
```diff
@@ -151,43 +198,43 @@ void Dialog::Draw()
...
 	SpriteShader::Draw(bottom, pos);
-	pos.Y() += bottom->Height() * .5 - 25.;
+	pos.Y() += (bottom->Height() - cancel->Height()) * .5;
 
 	// Draw the buttons, including optionally the cancel button.
 	const Color &bright = *GameData::Colors().Get("bright");
 	const Color &dim = *GameData::Colors().Get("medium");
 	const Color &back = *GameData::Colors().Get("faint");
 	const Color &inactive = *GameData::Colors().Get("inactive");
+	string okText = isMission ? "Accept" : "OK";
```
Comment (RisingLeaf) [inline source/Dialog.cpp:208]:
```suggestion
	const string okText = isMission ? "Accept" : "OK";
```

File: source/Dialog.cpp [hunk 2]
```diff
@@ -129,18 +177,17 @@ void Dialog::Draw()
 	const Sprite *bottom = SpriteSet::Get(isWide ? "ui/dialog bottom wide" : "ui/dialog bottom");
 	const Sprite *cancel = SpriteSet::Get("ui/dialog cancel");
 
-	// Get the position of the top of this dialog, and of the text and input.
-	Point pos(0., (top->Height() + height * middle->Height() + bottom->Height()) * -.5f);
-	Point textPos(Width() * -.5 + 10., pos.Y() + 20.);
-	Point inputPos = Point(0., -70.) - pos;
+	// Get the position of the top of this dialog, and of the input.
+	Point pos(0., (top->Height() + extensionCount * middle->Height() + bottom->Height()) * -.5f);
```
Comment (RisingLeaf) [inline source/Dialog.cpp:181]:
```suggestion
	Point pos(0., (top->Height() + extensionCount * middle->Height() + bottom->Height()) * -.5);
```
point takes in double, no idea why it was float here before

File: source/Dialog.cpp [hunk 3]
```diff
@@ -85,6 +82,57 @@ namespace {
 		{SDLK_KP_SPACE, ' '},
 		{SDLK_KP_VERTICALBAR, '|'}
 	};
+
+	// The width of the margin on the left side of the dialog. This area is part of the sprite,
+	// but shouldn't have any text or other graphics rendered over it. (It's mostly transparent.)
+	const int leftMargin = 20;
```
Comment (RisingLeaf) [inline source/Dialog.cpp:88]: why did you move away from CAPS_LOCK?

File: source/Dialog.cpp [hunk 4]
```diff
@@ -85,6 +82,57 @@ namespace {
...
+	// The overall padding width of an input field, in pixels.
+	const int inputHorizontalPadding = inputLeftPadding + inputRightPadding;
+	// The height of the padding at the top of an input field.
+	const int inputTopPadding = 2;
+	// The height of the padding at the bottom of an input field.
+	const int inputBottomPadding = 2;
+	// The overall padding height of an input field, in pixels.
+	const int inputVerticalPadding = inputTopPadding + inputBottomPadding;
+	// The height of an input field in pixels.
+	const int inputHeight = 20;
```
Comment (RisingLeaf) [inline source/Dialog.cpp:135]:
```suggestion
	// The width of the margin on the right/left sides of the dialog. This area is part of the sprite,
	// but shouldn't have any text or other graphics rendered over it. (It's mostly transparent.)
	const int leftMargin = 20;
	const int rightMargin = 20;
	const int horizontalMargin = leftMargin + rightMargin;
	// The margin on the right/left sides of the button sprite. The bottom segment also includes a button
	// that uses the same value.
	const int buttonLeftMargin = 10;
	const int buttonRightMargin = 10;
	const int buttonHorizontalMargin = buttonLeftMargin + buttonRightMargin;
	// The margin on the top/bottom sides of the button sprite. The bottom segment also includes a button
	// that uses the same value.
	const int buttonTopMargin = 10;
	const int buttonBottomMargin = 10;
	const int buttonVerticalMargin = buttonTopMargin + buttonBottomMargin;
	// The width of the padding used on the left/right sides of each segment, in pixels.
	const int leftPadding = 10;
	const int rightPadding = 10;
	const int horizontalPadding = rightPadding + leftPadding;
	// The height of the padding used by the top/bottom segment, in pixels.
	const int topPadding = 10;
	const int bottomPadding = 10;
	const int verticalPadding = topPadding + bottomPadding;
	// The width of the padding at the beginning/end of an input field.
	const int inputLeftPadding = 5;
	const int inputRightPadding = 5;
	const int inputHorizontalPadding = inputLeftPadding + inputRightPadding;
	// The height of the padding at the top/bottom of an input field.
	const int inputTopPadding = 2;
	const int inputBottomPadding = 2;
	const int inputVerticalPadding = inputTopPadding + inputBottomPadding;
	// The height of an input field in pixels.
	const int inputHeight = 20;
```
there is such a thing as too many comments where the readabillity decreases, I do expect the people who read the code have at least some capacity to  read variable names XD.

**Review by RisingLeaf — CHANGES_REQUESTED** [review]

**Comment by tibetiroka** [comment]
I don't want to use constexpr since we might want to define these values via interfaces in the future. I could change them to double, though.

**Comment by tibetiroka** [comment]
Using interfaces would be another refactor, since some of these constants are shared with conversations, for instance. (But they are only magic numbers for now.)

**Review by petervdmeer — APPROVED** [review]

**Pull request merged**